### PR TITLE
make audit report international

### PR DESCRIPTION
**Pull request by JiahuiZhao11**
Ⅰ. Describe what this PR does
add international util & make audit report international
Ⅱ. Does this pull request fix one issue?

Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.

Ⅳ. Describe how to verify it

Ⅴ. Special notes for reviews

### CODE REVIEW

**Comment by weilaaa** [comment]
better use multi instance to optimize it

File: pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func registerCubeAPI(cfg *Config) http.Handler {
 	apisOutsideMiddlewares(router)
 
 	// set middlewares for apis below
-	middlewares.SetUpMiddlewares(router)
+	middlewares.SetUpMiddlewares(router, cfg.Gi18nManagers)
```
Comment (weilaaa) [inline pkg/apiserver/apiserver.go:65]: set Gi18nManagers into cfg

File: pkg/apiserver/middlewares/audit/audit.go [hunk 2]
```diff
@@ -214,8 +213,12 @@ func handleProxyApi(ctx context.Context, c *gin.Context, e Event, t *gi18n.Manag
 	}
 
 	method := c.Request.Method
-	enT := en.GetEnManager()
+	enT := h.Managers.GetInstants("en")
```
Comment (weilaaa) [inline pkg/apiserver/middlewares/audit/audit.go:216]: consider if use const to cover string

File: pkg/apiserver/middlewares/audit/audit.go [hunk 3]
```diff
@@ -214,8 +213,12 @@ func handleProxyApi(ctx context.Context, c *gin.Context, e Event, t *gi18n.Manag
...
 	e.EventName = enT.Translate(ctx, method) + strings.Title(objectType[:len(objectType)-1])
+	t := h.Managers.GetInstants(env.AuditLanguage())
```
Comment (weilaaa) [inline pkg/apiserver/middlewares/audit/audit.go:218]: move it to handler init

File: pkg/apiserver/middlewares/audit/audit_test.go [hunk 4]
```diff
@@ -26,10 +27,13 @@ import (
 	"time"
 
 	"github.com/gin-gonic/gin"
+	"github.com/gogf/gf/v2/i18n/gi18n"
 	"github.com/google/uuid"
-	"github.com/kubecube-io/kubecube/pkg/utils/constants"
+	"github.com/kubecube-io/kubecube/pkg/utils/international"
```
Comment (weilaaa) [inline pkg/apiserver/middlewares/audit/audit_test.go:32]: move this pkg bellow

File: pkg/apiserver/middlewares/audit/audit_test.go [hunk 5]
```diff
@@ -69,35 +73,37 @@ func TestSendEvent(t *testing.T) {
...
-	router.GET("/api/v1/cube/proxy/clusters/:cluster/apis/apps/v1/namespaces/:namespace/statefulsets/:name", func(c *gin.Context) {
-		e = handleProxyApi(c, *e)
-		return
-	})
-	_ = performRequest(router, http.MethodGet, "/api/v1/cube/proxy/clusters/pivot-cluster/apis/apps/v1/namespaces/dev/statefulsets/stsA", []byte(""))
-	if e.EventName != "[KubeCube] query statefulsets" {
-		t.Fail()
+	ctx := context.Background()
+	enInstance := gi18n.Instance()
+	enInstance.SetPath("/Users/zhaojiahui/code/go/kubecube/KubeCube/i18n")
```
Comment (weilaaa) [inline pkg/apiserver/middlewares/audit/audit_test.go]: do not use absolute path here

File: pkg/utils/international/manager.go [hunk 6]
```diff
@@ -16,11 +16,72 @@ limitations under the License.
...
+	}
+	cm := &v1.ConfigMap{}
+	err := kClient.Get(context.Background(), client.ObjectKey{Name: "kubecube-language-config", Namespace: constants.CubeNamespace}, cm)
+	if err != nil && !ioerrors.IsNotFound(err) {
+		clog.Error("get configmap kubecube-language-config from K8s err: %v", err)
+		return nil, err
+	}
+	var languages []string
+	if ioerrors.IsNotFound(err) {
+		languages = defaultOptions
```
Comment (weilaaa) [inline pkg/utils/international/manager.go:58]: if configmap not found, i18n file less

File: pkg/utils/international/manager.go [hunk 7]
```diff
@@ -16,11 +16,72 @@ limitations under the License.
...
+	} else {
+		languages = strings.Split(cm.Data["languages"], ",")
+	}
+
+	m := &Gi18nManagers{
+		Managers: make(map[string]*gi18n.Manager),
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
```
Comment (weilaaa) [inline pkg/utils/international/manager.go:67]: newManager func do not need lock

File: pkg/utils/international/manager.go [hunk 8]
```diff
@@ -16,11 +16,72 @@ limitations under the License.
...
+
+func (g *Gi18nManagers) GetInstants(language string) *gi18n.Manager {
+	g.lock.RLock()
+	defer g.lock.RUnlock()
+	return g.Managers[language]
+}
+
+func (g *Gi18nManagers) Translate(ctx context.Context, language string, content string) string {
+	m := g.Managers[language]
+	g.lock.RLock()
```
Comment (weilaaa) [inline pkg/utils/international/manager.go:84]: put m := g.Managers[language] init lock state

**Review by weilaaa — CHANGES_REQUESTED** [review]
need fix something

File: pkg/apiserver/apiserver.go (hunk 1 above)
Comment (JiahuiZhao11) [reply pkg/apiserver/apiserver.go:65]: do not modify temporarily

File: pkg/apiserver/middlewares/audit/audit.go [hunk 9]
```diff
@@ -306,3 +315,25 @@ func (r responseBodyWriter) WriteString(s string) (n int, err error) {
...
+
+func getPostObjectName(c *gin.Context) string {
+	data, err := c.GetRawData()
+	body := make(map[string]interface{})
+	err = json.Unmarshal(data, &body)
+	if err != nil {
+		clog.Error("[audit] unmarshal request body err: %s", err.Error())
+		return ""
+	}
+	metadata, ok := body["metadata"].(map[string]interface{})
```
Comment (weilaaa) [inline pkg/apiserver/middlewares/audit/audit.go:327]: what if there is no key "metadata"

File: pkg/apiserver/middlewares/audit/audit.go [hunk 10]
```diff
@@ -306,3 +315,25 @@ func (r responseBodyWriter) WriteString(s string) (n int, err error) {
...
+	if !ok {
+		clog.Error("[audit] convert metadata to map failed")
+		return ""
+	}
+	c.Request.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+	name, ok := metadata["name"].(string)
```
Comment (weilaaa) [inline pkg/apiserver/middlewares/audit/audit.go:333]: as same as before

**Review by weilaaa — CHANGES_REQUESTED** [review]
need fix

**Review by weilaaa — APPROVED** [review]
LGTM

**Review by weilaaa — APPROVED** [review]
LGTM

**Pull request merged**